### PR TITLE
Refactored item-get

### PR DESCRIPTION
**Pull request by ovekyc**
* Added item-alive index promise
* Added item-value getting promise
* Added valid cheching promise

Refactoring item-get

There are a lot of callback in item-get function. I refactoring it like few depth callbacks using promises

### CODE REVIEW

**Comment by ovekyc** [comment]
Not complete yet

File: src/server/items/controllers.js [hunk 1]
```diff
@@ -39,7 +39,22 @@ export default {
...
+          // TODO : refactoring, It does not seem good accessing S3Connector directly in controller.
+          if (isThumbnail === 'true') {
+            item.imageUrls = s3Connector.getPrefixedImageUrls(images, IMAGE_SIZE_PREFIX.THUMBNAIL);
+          } else {
+            item.imageUrls = s3Connector.getImageUrls(images);
+          }
+          resolve(item);
+        });
+      }).then((valueList) => {
+        console.log(valueList);
```
Comment (ochanje210) [inline src/server/items/controllers.js]: then(console.log)

**Comment by ovekyc** [comment]
Refactoring complete, Please review it.

There is some wierd point in existed code.
If some key is not existed then our logic call reject.
There is fail-fast-behavior in Promise.all. Please check out link below.
https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise/all

And it means, The exception handling when notFound error occur, was not proper.
Fail-fast throws error right after meeting error. So further logic does not work and it also means last items didnt contain any item values.

So i just comment out that code block

**Comment by ochanje210** [comment]
understood. lack of our knowledge on fail-fast behavior of Promise.all resulted unexpected behavior from our server logics.

File: src/server/items/controllers.js [hunk 2]
```diff
@@ -13,72 +13,48 @@ export default {
...
+      // get item value
+      Promise.all(proms).then(lists =>
+        lists.reduce((result, list) => result.concat(list)))
+      .then(values => Promise.all(values.map(value => getPromise(value.key))))
+      // valid checking
+      .then(items => Promise.all(items.map(item => new Promise(resolve => {
+        ItemManager.validChecker(item, valid =>
+          valid ? resolve(item) : resolve(null));
+      }))))
+      .then(items => items.filter((item) => item !== null))
```
Comment (hewonjeong) [inline src/server/items/controllers.js]: how about `.then(items => items.filter(item => item))`?

Comment (ochanje210) [reply src/server/items/controllers.js]: are those same logic?

File: src/server/items/controllers.js [hunk 3]
```diff
@@ -13,72 +13,48 @@ export default {
...
-      .then(() => {
-        res.status(200).send({
-          items
+
+      // get item-alive (index)
+      const proms = keys.map(key => new Promise((resolve, reject) => {
+        const itemIndexKey = `${ENTITY.ITEM}-${STATE.ALIVE}-${key}-`;
+        fetchPrefix(itemIndexKey, (err, list) =>
+          err ? reject(err) : resolve(list));
+      }));
```
Comment (hewonjeong) [inline src/server/items/controllers.js]: any reason you defined `proms` variable?
usually `Promise.all(keys.map(key => new Promise((resolve, reject) => ~~~~))` is more familiar with our code

**Review by festinalente91 — APPROVED** [review]

File: src/server/items/controllers.js (hunk 2 above)
Comment (hewonjeong) [reply src/server/items/controllers.js]: @ochanje210 cause item value is null or json object, so I think we can filter with just `=> item`

Comment (ovekyc) [reply src/server/items/controllers.js]: What is `item => item`?
I thought like filter should return true or false. 
Even if when it return null, it maybe work. but i think `item !== null` has more readability.

File: src/server/items/controllers.js [hunk 4]
```diff
@@ -16,14 +16,13 @@ export default {
       const s3Connector = new S3Connector();
 
       // get item-alive (index)
-      const proms = keys.map(key => new Promise((resolve, reject) => {
+      Promise.all(keys.map(key => new Promise((resolve, reject) => {
         const itemIndexKey = `${ENTITY.ITEM}-${STATE.ALIVE}-${key}-`;
         fetchPrefix(itemIndexKey, (err, list) =>
           err ? reject(err) : resolve(list));
-      }));
+      })))
```
Comment (ochanje210) [inline src/server/items/controllers.js]: don't u still need semi-colon?

File: src/server/items/controllers.js (hunk 2 above)
Comment (hewonjeong) [reply src/server/items/controllers.js]: @ovekyc OK. I think so.:)

**Pull request merged**